# **APCO 1P93: Applied Programming (for Data Science)**

### Winter, 2020
### Instructor: Yifeng (Ethan) Li
### Department of Computer Science, Brock University
### Email: <yli2@brocku.ca>
---

**Lanaguage:** Python 3.6, __3.7__, or 3.8

**Textbooks and materials:**
* John Zelle, Python Programming: An Introduction to Computer Science, 3rd Ed. Franklin, Beedle & Associates Inc. 2016.  
* Python 3.7.6 documentation: <https://docs.python.org/3.7>
* Packages' official documentations
* Online materials
* My experience
---

## **Outline**
1. **Basics**
  1. Platforms, IEDs, Jupyter Notebook/Lab.
  2. Basics (data types, int, float, string, list, dictionary, tuple, set, etc)
  3. conditional, loop, control statements
  4. functions
  5. _Assignment (Jupyter Lab), Due on January 23, 2020_
2. **Modules**
  1. Numeric and mathematical modules
  2. Creat your own modules
3. **Class and Object-Oriented Programming**
  1. _Assignment_
4. **File IO**
  1. IO for sequential files
  2. IO for binary files
  3. Data Persistence
  4. Data Compression and Archiving
  5. File Formats
  5. _Assignment (sequential file and binary file)_
5. **Python for Scientific Computing**
  1. SciPy: Python-based ecosystem of open-source software for mathematics, science, and engineering
  2. NumPy for scientific computing with Python
  3. pandas for data structures & analysis
  4. _Assignment (NMF)_
6. **Python for Graphics and Data Visualization**
  1. GGraphical User Interfaces with Tkinter (Tk)
  2. matplotlib: Python plotting library
  3. seaborn: statistical data visualization
  4. _Assignment (data analytics and visualization)_
7. **Errors and Exceptions**
8. **Searching, Sorting and Recursion**
  1. Search
  2. Sort
  3. _Assignment_
8. **Software packing and distribution**
  1. Software packing
  2. Calling functions from C/C++
  3. Github
  4. Anaconda Cloud
  5. Virtual environment and Docker
  

## **Homeworks and Tests**
* **Attendence:** 5%
* **Homeworks:** 6, one homework per two weeks, submit in lab sessions. (5% * 6 = 30%)
* **Midterm exam:** 1.5 or 3 hours. (30%)
* **Final exam:** 3 hours. (35%)



## **Python Package Management Using Anaconda or pip**
* **What is Anaconda?**
Anaconda is a free and open-source distribution of the Python and R programming languages for scientific computing (data science, machine learning applications, large-scale data processing, predictive analytics, etc.), that aims to simplify package management and deployment. Package versions are managed by the package management system _conda_. The Anaconda distribution includes data-science packages suitable for Windows, Linux, and MacOS.

* [Anaconda distribution](https://www.anaconda.com) comes with more than 1,500 packages as well as the conda package and virtual environment manager. It also includes a GUI, Anaconda Navigator, as a graphical alternative to the command line interface (CLI).

* **What is Anaconda cloud?**
[Anaconda Cloud](https://anaconda.org) is a package management service by Anaconda where you can find, access, store and share public and private notebooks, environments, and conda and PyPI packages. Cloud hosts useful Python packages, notebooks and environments for a wide variety of applications. You do not need to log in or to have a Cloud account, to search for public packages, download and install them.

* You can build new packages using the Anaconda Client command line interface (CLI), then manually or automatically upload the packages to Cloud.

* **What is Conda?**
Conda is an open source, cross-platform, language-agnostic package manager and environment management system that installs, runs, and updates packages and their dependencies. It was created for Python programs, but it can package and distribute software for any language (e.g., R), including multi-language projects. The conda package and environment manager is included in all versions of Anaconda, Miniconda, and Anaconda Repository.

* **What is pip?**
[pip](https://pypi.org/project/pip) is the package installer for Python. You can use pip to install packages from the Python Package Index and other indexes. Most distributions of Python come with pip preinstalled. Python 2.7.9 and later (on the python2 series), and Python 3.4 and later include pip (pip3 for Python 3) by default.

* **Difference between Anaconda and pip?**
  * The big difference between conda and the pip package manager is in how package dependencies are managed, which is a significant challenge for Python data science and the reason conda was created. pip installs all Python package dependencies required, whether or not those conflict with other packages previously installed. So a working installation of, for example, Google TensorFlow can suddenly stop working when you pip-install a new package that needs a different version of the NumPy library. 
  * More insidiously, everything might still appear to work, but now you get different results, or you are unable to reproduce the same results elsewhere because you didn't pip-install in the same order.



* **How to install Anaconda?**
  * Download `Anaconda3-2019.10-Linux-x86_64.sh` from <https://www.anaconda.com/distribution/#download-section>. 
  * `bash ~/Downloads/Anaconda3-2019.10-Linux-x86_64.sh`

* **How to use Anaconda?**
  * Install packages:
    * `conda install <PKGNAME>`
    * `conda install -c conda-forge jupyterlab`
    * `conda remove <PKGNAME1> <PKGNAME2> ...
    * `conda remove scipy curl`
  * Update conda:
    * `conda update conda`

* **More information about Anaconda and conda**
  * [Anaconda Documentation](https://docs.anaconda.com/anaconda)
  * [Conda Documentation](https://conda.io/projects/conda/en/latest/index.html)
  * [Conda Cheat Sheet](https://conda.io/projects/conda/en/latest/user-guide/cheatsheet.html)

## **Jupyter Lab**
* What is Jupyter Lab?
  * JupyterLab is a next-generation web-based user interface for Project Jupyter.
  * <https://jupyterlab.readthedocs.io/en/stable>
* Install and uninstall.
  * `conda install -c conda-forge jupyterlab`
  * `conda remove jupyterlab`
* We will mainly use Jupyter Lab for lectures and homeworks.
* Running Python code on Jupyter Lab.
* Markdown Language.

## **IDE**
1. **Spyder** (https://www.spyder-ide.org)
2. PyCharm (https://www.jetbrains.com/pycharm)
3. PyDev (https://www.pydev.org)

We may use Spyder.

---

## **Basic Python Data Types**

In [ ]:
print("Hello Word!")

### **Comments**

Comments in Python start with the hash character, #, and extend to the end of the physical line. A comment may appear at the start of a line or following whitespace or code, but not within a string literal. A hash character within a string literal is just a hash character. Since comments are to clarify code and are not interpreted by Python, they may be omitted when typing in examples.

In [ ]:
a=1 # assign 1 to first addend
b=2 # assign 2 to second addend
c=a+b # add a and b together
print(c)
text = "# This is not a comment because it's inside quotes."
print(text)

### **Numbers**

The interpreter acts as a simple calculator: you can type an expression at it and it will write the value. Expression syntax is straightforward: the operators +, -, * and / work just like in most other languages (for example, Pascal or C); parentheses (()) can be used for grouping.

In [ ]:
(2+3)*4

In [ ]:
(2+3)*4.0

The integer numbers (e.g. 2, 3, 4, 20) have type **int**, the ones with a fractional part (e.g. 4.0, 20.0) have type **float**. 

Division (/) always returns a float. To do floor division and get an integer result (discarding any fractional result) you can use the // operator; to calculate the remainder you can use %.

In [ ]:
20/3

In [6]:
20//3

6

In [7]:
20%3

2

With Python, it is possible to use the ** operator to calculate powers.

In [8]:
2**3

8

The equal sign (=) is used to assign a value to a variable. Afterwards, no result is displayed before the next interactive prompt:

In [9]:
width = 20
height = 5 * 9
width * height

900

In [10]:
area = width * height
area
print('AREA:')
print(area)

AREA:
900


If a variable is not “defined” (assigned a value), trying to use it will give you an error:

In [11]:
#areas = areas + area

There is full support for floating point; operators with mixed type operands convert the integer operand to floating point:

In [12]:
4 * 3.75 - 1

14.0

In interactive mode, the last printed expression is assigned to the variable _. This means that when you are using Python as a desk calculator, it is somewhat easier to continue calculations, for example:

In [13]:
tax = 13 / 100
price = 100.50
price * tax

13.065000000000001

In [14]:
price + _

113.565

In [15]:
round(_, 2)

113.56

This variable should be treated as read-only by the user. Don’t explicitly assign a value to it — you would create an independent local variable with the same name masking the built-in variable with its magic behavior.

In [16]:
a=3
b=4
_=a+b
c=_ + 3
print(c)
print(_)

10
7


Convert between `int` and `float` using `int( )` and `float( )`:

In [17]:
a = 3
print(float(a))

3.0


In [18]:
b = 3.5
print(int(b))

3


In addition to int and float, Python supports other types of numbers, such as Decimal and Fraction. Python also has built-in support for complex numbers, and uses the j or J suffix to indicate the imaginary part (e.g. 3+5j).

### **Strings**

Besides numbers, Python can also manipulate strings, which can be expressed in several ways. They can be enclosed in single quotes (`'...'`) or double quotes (`"..."`) with the same result 2. \ can be used to escape quotes:

In [19]:
'Brock University'

'Brock University'

In [20]:
"Computer Science"

'Computer Science'

In [21]:
'"Yes" they said.'

'"Yes" they said.'

In [22]:
"\"Yes,\" they said."

'"Yes," they said.'

In [23]:
'"Isn\'t," they said.'

'"Isn\'t," they said.'

In the interactive interpreter, the output string is enclosed in quotes and special characters are escaped with backslashes. While this might sometimes look different from the input (the enclosing quotes could change), the two strings are equivalent. The string is enclosed in double quotes if the string contains a single quote and no double quotes, otherwise it is enclosed in single quotes. The print() function produces a more readable output, by omitting the enclosing quotes and by printing escaped and special characters:

In [24]:
print('"Isn\'t," they said.')

"Isn't," they said.


In [25]:
s = 'First line.\nSecond line.'  # \n means newline
s # without print(), \n is included in the output

'First line.\nSecond line.'

In [26]:
print(s)  # with print(), \n produces a new line

First line.
Second line.


If you don’t want characters prefaced by \ to be interpreted as special characters, you can use raw strings by adding an r before the first quote:

In [27]:
print('C:\some\name')  # here \n means newline!
print(r'C:\some\name')  # note the r before the quote

C:\some
ame
C:\some\name


String literals can span multiple lines. One way is using triple-quotes: `"""..."""` or `'''...'''`. End of lines are automatically included in the string, but it’s possible to prevent this by adding a \ at the end of the line. The following example:

In [28]:
print("""\
Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to
""")

Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to



Strings can be concatenated (glued together) with the `+` operator, and repeated with `*`:

In [29]:
3 *'> ' +  ' Brock ' + 'University'

'> > >  Brock University'

Two or more string literals (i.e. the ones enclosed between quotes) next to each other are automatically concatenated.

In [30]:
'Py' 'th' 'on' + '3.7'

'Python3.7'

This feature is particularly useful when you want to break long strings:

In [31]:
text = ('Put several strings within parentheses '
         'to have them joined together.')
text

'Put several strings within parentheses to have them joined together.'

This only works with two literals though, not with variables or expressions:

In [32]:
prefix = 'Py'
#prefix 'thon'  # can't concatenate a variable and a string literal

If you want to concatenate variables or a variable and a literal, use `+`:

In [33]:
prefix = 'Py'
prefix + 'thon'

'Python'

Strings can be indexed (subscripted), with the first character having index 0. There is no separate character type; a character is simply a string of size one:

In [34]:
word = 'Python'
print(word[0])  # character in position 0
print(word[5])  # character in position 5

P
n


Indices may also be negative numbers, to start counting from the right:

In [35]:
print(word[-1])  # last character
print(word[-2])  # second-last character
print(word[-6])

n
o
P


Note that since -0 is the same as 0, negative indices start from -1.

In addition to indexing, _slicing_ is also supported. While indexing is used to obtain individual characters, slicing allows you to obtain substring:

In [36]:
print(word[0:2])  # characters from position 0 (included) to 2 (excluded)
print(word[2:5])  # characters from position 2 (included) to 5 (excluded)

Py
tho


Note how the start is always included, and the end always excluded. This makes sure that `s[:i] + s[i:]` is always equal to `s`:

In [37]:
print(word[:2] + word[2:])
print(word[:4] + word[4:])

Python
Python


Slice indices have useful defaults; an omitted first index defaults to zero, an omitted second index defaults to the size of the string being sliced.

In [38]:
word[:2]   # character from the beginning to position 2 (excluded)

'Py'

In [39]:
word[4:]

'on'

In [40]:
word[-2:]  # characters from the second-last (included) to the end

'on'

Attempting to use an index that is too large will result in an error:

In [41]:
#word[42]  # the word only has 6 characters

However, out of range slice indexes are handled gracefully when used for _slicing_:

In [42]:
word[4:42]

'on'

In [43]:
word[42:]

''

Python strings cannot be changed — they are **immutable**. Therefore, assigning to an indexed position in the string results in an error:

In [44]:
#word[0] = 'J'

In [45]:
#word[2:] = 'py'

If you need a different string, you should create a new one:

In [46]:
'C' + word[1:]

'Cython'

In [47]:
word[:2] + 'py'

'Pypy'

In [48]:
import numpy as np
wordnp= np.array(['Python','Language'],dtype=str)
print(wordnp)

['Python' 'Language']


In [49]:
wordnp
print(wordnp[0])
print(wordnp[0][0])
wordnp[0] = 'C' + wordnp[0]
print(wordnp)

Python
P
['CPython' 'Language']


The built-in function `len()` returns the length of a string:

In [50]:
len(word)

6

More about print and strings will be taught in subsequent sessions.

### **Boolean Type**

Boolean type important for conditional flow control.

In [51]:
a = False
b = True
c = 1
c<0

False

In [52]:
a or b

True

In [53]:
a and b

False

In [54]:
not a

True

In [55]:
c is 1

True

In [56]:
c is not 0

True

In [57]:
# convert numbers to boolean type
c = 10
bool(c)

True

In [58]:
c = 0
bool(c)

False

In [59]:
c = -1
bool(c)

True

In [60]:
# convert boolean type to numbers
int(True)

1

In [61]:
int(False)

0

### **Lists**

Python knows a number of compound data types, used to group together other values. The most versatile is the list, which can be written as a list of comma-separated values (items) between square brackets. Lists might contain items of _different types_, but usually the items all have the same type.

In [62]:
squares = [1, 4, 9, 16, 25]
squares

[1, 4, 9, 16, 25]

Like strings (and all other built-in sequence types), lists can be **indexed** and **sliced**:

In [63]:
squares[0]  # indexing returns the item

1

In [64]:
squares[-1]

25

In [65]:
squares[-3:]

[9, 16, 25]

All slice operations return a new list containing the requested elements. This means that the following slice returns a _new_ (_shallow_) copy of the list:

In [66]:
squares[:]

[1, 4, 9, 16, 25]

In [67]:
sq2=squares[:4]
print(sq2)

[1, 4, 9, 16]


In [68]:
sq2[0]=0
sq2

[0, 4, 9, 16]

In [69]:
squares

[1, 4, 9, 16, 25]

In [70]:
sq3=squares
sq3

[1, 4, 9, 16, 25]

In [71]:
sq3[1]=3
sq3.append(36)
sq3

[1, 3, 9, 16, 25, 36]

In [72]:
squares

[1, 3, 9, 16, 25, 36]

Lists also support operations like concatenation:

In [73]:
squares + [49, 64, 81, 100]

[1, 3, 9, 16, 25, 36, 49, 64, 81, 100]

Unlike strings, which are immutable, lists are a mutable type, i.e. it is possible to change their content:

In [74]:
cubes = [1, 8, 27, 65, 125]  # something's wrong here

In [75]:
cubes[3] = 64  # replace the wrong value
cubes

[1, 8, 27, 64, 125]

You can also add new items at the end of the list, by using the `append()` method (we will see more about methods later):

In [76]:
cubes.append(216)  # add the cube of 6
cubes.append(7 ** 3)  # and the cube of 7
cubes

[1, 8, 27, 64, 125, 216, 343]

Assignment to slices is also possible, and this can even change the size of the list or clear it entirely:

In [77]:
letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g']
letters

['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [78]:
letters[2:5] = ['C', 'D', 'E'] # # replace some values
letters

['a', 'b', 'C', 'D', 'E', 'f', 'g']

In [79]:
letters[2:5] = [] # now remove them
letters

['a', 'b', 'f', 'g']

In [80]:
letters[:] = [] # clear the list by replacing all the elements with an empty list
letters

[]

The built-in function `len()` also applies to lists:

In [81]:
letters = ['a', 'b', 'c', 'd']
len(letters)

4

It is possible to nest lists (create lists containing other lists), for example:

In [82]:
a = ['a', 'b', 'c']
n = [1, 2, 3]
x = [a, n]
s = 5
[a,n,x,s]

[['a', 'b', 'c'], [1, 2, 3], [['a', 'b', 'c'], [1, 2, 3]], 5]

### **Tuples and Sequences**

We saw that lists and strings have many common properties, such as _indexing_ and _slicing_ operations. They are two examples of sequence data types (Sequence Types — list, tuple, range). Since Python is an evolving language, other sequence data types may be added. There is also another standard sequence data type: the tuple.

A tuple consists of a number of values separated by commas, for instance:

In [83]:
t = 12345, 54321, 'hello!' # packing
t

(12345, 54321, 'hello!')

In [84]:
t2 = (12, 34, 'abc')
t2

(12, 34, 'abc')

In [85]:
# Tuples may be nested:
u = t, (1, 2, 3, 4, 5)
u

((12345, 54321, 'hello!'), (1, 2, 3, 4, 5))

In [86]:
# Tuples are immutable:
#t[0] = 88888

In [87]:
# but they can contain mutable objects:
v = ([1, 2, 3], [3, 2, 1])
v[0][0]=0
v

([0, 2, 3], [3, 2, 1])

As you see, on output tuples are always enclosed in parentheses, so that nested tuples are interpreted correctly; they may be input **with or without surrounding parentheses**, although often parentheses are necessary anyway (if the tuple is part of a larger expression). It is not possible to assign to the individual items of a tuple, however it is possible to create tuples which contain mutable objects, such as lists.

Though tuples may seem similar to lists, they are often used in different situations and for different purposes. Tuples are **immutable**, and usually contain a **heterogeneous** sequence of elements that are accessed via **unpacking** (see later in this section) or **indexing** (or even by attribute in the case of namedtuples). Lists are **mutable**, and their elements are usually **homogeneous** and are accessed by **iterating** over the list.

In [88]:
# unpacking
v0,v1=v 
v0

[0, 2, 3]

In [89]:
# indexing
t3=(0, 11, 22, 33)
t4=t3[0:3]
print(t4)

for v in t3:
    print(v)

(0, 11, 22)
0
11
22
33


A special problem is the construction of tuples containing 0 or 1 items: the syntax has some extra quirks to accommodate these. Empty tuples are constructed by an empty pair of parentheses; a tuple with one item is constructed by following a value with a comma (it is not sufficient to enclose a single value in parentheses). Ugly, but effective. For example:

In [90]:
empty = ()
len(empty)

0

In [91]:
singleton = 'hello',    # <-- note trailing comma
len(singleton)

1

In [92]:
singleton2 = ('hello',)
singleton2

('hello',)

The statement `t = 12345, 54321, 'hello!'` is an example of tuple packing: the values `12345`, `54321` and `'hello!'` are packed together in a tuple. The reverse operation is also possible:

In [93]:
x, y, z = t # unpacking
x

12345

This is called, appropriately enough, sequence unpacking and works for any sequence on the right-hand side. Sequence unpacking requires that there are as many variables on the left side of the equals sign as there are elements in the sequence. Note that multiple assignment is really just a combination of tuple packing and sequence unpacking.

### **Dictionaries**

Another useful data type built into Python is the dictionary (see Mapping Types — dict). Dictionaries are sometimes found in other languages as “associative memories” or “associative arrays”. Unlike sequences, which are indexed by a range of numbers, dictionaries are indexed by keys, which can be **any immutable type; _strings_ and _numbers_ can always be keys**. Tuples can be used as keys if they contain only strings, numbers, or tuples; if a tuple contains any mutable object either directly or indirectly, it cannot be used as a key. You can’t use lists as keys, since lists can be modified in place using index assignments, slice assignments, or methods like `append()` and `extend()`.

It is best to think of a dictionary as **a set of key:value pairs**, with the requirement that the keys are _unique_ (within one dictionary). A pair of braces creates an empty dictionary: `{}`. Placing a comma-separated list of key:value pairs within the braces adds initial key:value pairs to the dictionary; this is also the way dictionaries are written on output.

In [94]:
# empty dict
d={}
d

{}

In [95]:
# create a dict
d={'St. Catharines':133.1, 'NOTL':17.5, 'Niagara Falls':88.1, 'Welland':52.3, 
   ('Grimsby','Thorold') : [27.3,18.8], 
   'others':{'Lincoln':23.7, 'West Lincoln':14.5, 'Wainfleet':6.4, 'Pelham':17.1, 'Fort Erie':30.7},
   905:'area code',
  'Toronto':2930}
d2=dict(d) # copy
d

{'St. Catharines': 133.1,
 'NOTL': 17.5,
 'Niagara Falls': 88.1,
 'Welland': 52.3,
 ('Grimsby', 'Thorold'): [27.3, 18.8],
 'others': {'Lincoln': 23.7,
  'West Lincoln': 14.5,
  'Wainfleet': 6.4,
  'Pelham': 17.1,
  'Fort Erie': 30.7},
 905: 'area code',
 'Toronto': 2930}

The main operations on a dictionary are storing a value with some key and extracting the value given the key. It is also possible to delete a key:value pair with `del`. If you store using a key that is already in use, the old value associated with that key is _forgotten_. It is an error to extract a value using a non-existent key.

In [96]:
# access
d['St. Catharines']

133.1

In [97]:
d[905]

'area code'

In [98]:
d[('Grimsby', 'Thorold')]

[27.3, 18.8]

In [99]:
# delete
del d['Toronto']
d

{'St. Catharines': 133.1,
 'NOTL': 17.5,
 'Niagara Falls': 88.1,
 'Welland': 52.3,
 ('Grimsby', 'Thorold'): [27.3, 18.8],
 'others': {'Lincoln': 23.7,
  'West Lincoln': 14.5,
  'Wainfleet': 6.4,
  'Pelham': 17.1,
  'Fort Erie': 30.7},
 905: 'area code'}

In [100]:
# add
d['Niagara Region'] = 447.9
d

{'St. Catharines': 133.1,
 'NOTL': 17.5,
 'Niagara Falls': 88.1,
 'Welland': 52.3,
 ('Grimsby', 'Thorold'): [27.3, 18.8],
 'others': {'Lincoln': 23.7,
  'West Lincoln': 14.5,
  'Wainfleet': 6.4,
  'Pelham': 17.1,
  'Fort Erie': 30.7},
 905: 'area code',
 'Niagara Region': 447.9}

Performing `list(d)` on a dictionary returns a list of all the keys used in the dictionary, in insertion order (if you want it sorted, just use `sorted(d)` instead). To check whether a single key is in the dictionary, use the `in` keyword.

In [101]:
list(d)

['St. Catharines',
 'NOTL',
 'Niagara Falls',
 'Welland',
 ('Grimsby', 'Thorold'),
 'others',
 905,
 'Niagara Region']

In [102]:
#sorted(d) # error

In [103]:
del d[('Grimsby','Thorold')]
d

{'St. Catharines': 133.1,
 'NOTL': 17.5,
 'Niagara Falls': 88.1,
 'Welland': 52.3,
 'others': {'Lincoln': 23.7,
  'West Lincoln': 14.5,
  'Wainfleet': 6.4,
  'Pelham': 17.1,
  'Fort Erie': 30.7},
 905: 'area code',
 'Niagara Region': 447.9}

In [104]:
# sorted(d) error

In [105]:
del d[905]
sorted(d)

['NOTL',
 'Niagara Falls',
 'Niagara Region',
 'St. Catharines',
 'Welland',
 'others']

In [106]:
# logical
'NOTL' in d

True

In [107]:
'Toronto' not in d

True

The `dict()` constructor builds dictionaries directly from sequences of key-value pairs:

In [108]:
# creat dictionary
pc=dict( [('Ontario','Toronto'), ('Qubec', 'Qubec City'), ('British Columbia','Victoria')] )
pc

{'Ontario': 'Toronto', 'Qubec': 'Qubec City', 'British Columbia': 'Victoria'}

In addition, **dict comprehensions** can be used to create dictionaries from arbitrary key and value expressions:

In [109]:
# creat dictionary
sq={x: x**2 for x in (1, 2, 3, 4)}
sq

{1: 1, 2: 4, 3: 9, 4: 16}

In [110]:
# x:function(x)
import math
sq={x: math.exp(x) for x in (1, 2, 3, 4)}
sq

{1: 2.718281828459045,
 2: 7.38905609893065,
 3: 20.085536923187668,
 4: 54.598150033144236}

When the keys are simple strings, it is sometimes easier to specify pairs using keyword arguments:

In [111]:
# creat dictionary
tc=dict(Northwest_Territories = 'Yellowknife',
       Yukon = 'Whitehorse',
       Nunavut = 'Iqaluit')
tc

{'Northwest_Territories': 'Yellowknife',
 'Yukon': 'Whitehorse',
 'Nunavut': 'Iqaluit'}

### **Sets**

Python also includes a data type for sets. A set is an unordered collection with no duplicate elements. Basic uses include membership testing and eliminating duplicate entries. Set objects also support mathematical operations like union, intersection, difference, and symmetric difference.

Curly braces or the `set()` function can be used to create sets. Note: to create an empty set you have to use `set()`, not `{}`; the latter creates an empty dictionary, a data structure that we discuss in the next section.

Here is a brief demonstration:

In [112]:
basket = {'apple', 'orange', 'apple', 'pear', 'orange', 'banana'}
print(basket) # show that duplicates have been removed

{'pear', 'apple', 'banana', 'orange'}


In [113]:
'orange' in basket # fast membership testing

True

In [114]:
'crabgrass' in basket

False

In [115]:
# Demonstrate set operations on unique letters from two words
a = set('abracadabra')
b = set('alacazam')
c = set(('aab', 'ccdd','eefgg'))
print(a) # unique letters in a
print(b) # unique letters in b
print(c)

{'d', 'r', 'c', 'a', 'b'}
{'z', 'm', 'l', 'c', 'a'}
{'eefgg', 'aab', 'ccdd'}


In [116]:
a - b # letters in a but not in b

{'b', 'd', 'r'}

In [117]:
a | b # letters in a or b or both

{'a', 'b', 'c', 'd', 'l', 'm', 'r', 'z'}

In [118]:
a & b # letters in both a and b

{'a', 'c'}

In [119]:
a ^ b # letters in a or b but not both

{'b', 'd', 'l', 'm', 'r', 'z'}

Similarly to **list comprehensions**, **set comprehensions** are also supported:

In [120]:
a = {x for x in 'abracadabra' if x not in 'abc'}
a

{'d', 'r'}

---

## **Control Flow Tools**

### **if Statement**

Perhaps the most well-known statement type is the if statement. For example:

In [121]:
#x = int(input("Please enter an integer: "))
x=10

In [122]:
if x < 0:
    x = 0
    print('Negative changed to zero')
elif x == 0:
    print('Zero')
elif x == 1:
    print('Single')
else:
    print('More')

More


There can be zero or more elif parts, and the else part is optional. The keyword `elif` is short for `else if`, and is useful to avoid excessive indentation. An `if … elif … elif …` sequence is a substitute for the `switch` or `case` statements found in other languages.

In [123]:
#x = input("Please input save/delete/create: ")
x='save'

In [124]:
if x == 'save':
    print('The file is saved.')
elif x == 'delete':
    confirm = bool(input('Do you really want to delete this file?: True/False'))
    if confirm == True:
        print('The file is deleted!')
    else:
        print('The file is not deleted...')
elif x == 'create':
    print('A new file is created.')
else:
    print('Function not defined!')

The file is saved.


### **One line if statement in Python (ternary conditional operator)**

`<expression1> if <condition> else <expression2>`

It first evaluates the condition; if it returns `True`, `expression1` will be evaluated to give the result, otherwise `expression2`. Evaluation is lazy, so only one expression will be executed.

In [125]:
#weather = input('Weather: good or bad?')
weather = 'bad'
decision = 'work from home' if weather=='bad' else 'go to the university'
decision

'work from home'

### **for Statement**

The `for` statement in Python differs a bit from what you may be used to in C or Pascal. Rather than always iterating over an arithmetic progression of numbers (like in Pascal), or giving the user the ability to define both the iteration step and halting condition (as C), Python’s `for` statement iterates over the items of any sequence (a list or a string), in the order that they appear in the sequence. For example (no pun intended):

In [126]:
# Measure some strings:
words = ['cat', 'window', 'defenestrate']
for w in words:
    print(w, len(w))

cat 3
window 6
defenestrate 12


In [127]:
d={'St. Catharines':133.1, 'NOTL':17.5, 'Niagara Falls':88.1, 'Welland':52.3, 
   ('Grimsby','Thorold') : [27.3,18.8], 
   'others':{'Lincoln':23.7, 'West Lincoln':14.5, 'Wainfleet':6.4, 'Pelham':17.1, 'Fort Erie':30.7},
   905:'area code',
  'Toronto':2930}
for c in d:
    print(c,'--->' ,d[c])

St. Catharines ---> 133.1
NOTL ---> 17.5
Niagara Falls ---> 88.1
Welland ---> 52.3
('Grimsby', 'Thorold') ---> [27.3, 18.8]
others ---> {'Lincoln': 23.7, 'West Lincoln': 14.5, 'Wainfleet': 6.4, 'Pelham': 17.1, 'Fort Erie': 30.7}
905 ---> area code
Toronto ---> 2930


If you need to modify the sequence you are iterating over while inside the loop (for example to duplicate selected items), it is recommended that you first **make a copy**. Iterating over a sequence does not implicitly make a copy. The **slice** notation makes this especially convenient:

In [128]:
for w in words[:]:  # Loop over a slice copy of the entire list.
    if len(w) > 6:
        words.insert(0, w)
words

['defenestrate', 'cat', 'window', 'defenestrate']

With `for w in words:`, the example would attempt to create an infinite list, inserting defenestrate over and over again.

### **The range( ) Function**

If you do need to iterate over a sequence of numbers, the built-in function `range()` comes in handy. It generates arithmetic progressions:

In [129]:
import math
for i in range(10):
    print(math.log(i+1))

0.0
0.6931471805599453
1.0986122886681098
1.3862943611198906
1.6094379124341003
1.791759469228055
1.9459101490553132
2.0794415416798357
2.1972245773362196
2.302585092994046


The given end point is never part of the generated sequence; `range(10)` generates 10 values, the legal indices for items of a sequence of length 10. It is possible to let the range start at another number, or to specify a different increment (even negative; sometimes this is called the ‘step’):

In [130]:
for i in range(0,10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [131]:
list(range(2,10,2))

[2, 4, 6, 8]

In [132]:
list(range(10,-10,-2))

[10, 8, 6, 4, 2, 0, -2, -4, -6, -8]

In many ways the object returned by `range()` behaves as if it is a list, but in fact it isn’t. It is an object which returns the successive items of the desired sequence when you iterate over it, but it doesn’t really make the list, thus **saving space**.

We say such an object is iterable, that is, suitable as a target for functions and constructs that expect something from which they can obtain successive items until the supply is exhausted. We have seen that the `for` statement is such an iterator. The function `list()` is another; it creates lists from iterables.

Later we will see more functions that return iterables and take iterables as argument.

To iterate over the indices of a sequence, you can combine `range()` and `len()` as follows:

In [133]:
# a = ['Mary', 'had', 'a', 'little', 'lamb']
a = ('Mary', 'had', 'a', 'little', 'lamb')
# a = {'Mary', 'had', 'a', 'little', 'lamb'} # TypeError: 'set' object is not subscriptable
for i in range(len(a)):
    print(i, a[i])

0 Mary
1 had
2 a
3 little
4 lamb


### **break and continue Statements, and else Clauses on Loops**

The `break` statement, like in C, breaks out of the innermost enclosing `for` or `while` loop.

Loop statements may have an `else` clause; it is executed when the loop terminates through exhaustion of the list (with `for`) or when the condition becomes false (with `while`), but not when the loop is terminated by a `break` statement. This is exemplified by the following loop, which searches for prime numbers:

In [134]:
# else in for
for i in range(10):
    print(i)
else:
    print('out of boundary, this is the end')
    print(i)
        

0
1
2
3
4
5
6
7
8
9
out of boundary, this is the end
9


In [135]:
# else in while
i=2
while i<10:
    print(i)
    i=i+2
else:
    print('out of boundary, this is the end')
    print(i)

2
4
6
8
out of boundary, this is the end
10


In [136]:
# break and else
for n in range(2, 10):
    for x in range(2, n):
        if n % x == 0:
            print(n, 'equals', x, '*', n//x)
            break
    else:
        # loop fell through without finding a factor
        print(n, 'is a prime number')

2 is a prime number
3 is a prime number
4 equals 2 * 2
5 is a prime number
6 equals 2 * 3
7 is a prime number
8 equals 2 * 4
9 equals 3 * 3


When used with a loop, the `else` clause has more in common with the `else` clause of a `try` statement than it does that of `if` statements: a `try` statement’s `else` clause runs when no exception occurs, and a loop’s `else` clause runs when no `break` occurs. For more on the `try` statement and exceptions, see *Handling Exceptions*.

The `continue` statement, also borrowed from C, continues with the next iteration of the loop:

In [137]:
for num in range(2, 10):
    if num % 2 == 0:
        print("Found an even number", num)
        continue
    print("Found a number", num)

Found an even number 2
Found a number 3
Found an even number 4
Found a number 5
Found an even number 6
Found a number 7
Found an even number 8
Found a number 9


### **pass Statements**

The `pass` statement does nothing. It can be used when a statement is required syntactically but the program requires no action. For example:

In [138]:
#while True:
#    pass  # Busy-wait for keyboard interrupt (Ctrl+C)

This is commonly used for creating minimal classes:

In [139]:
class MyEmptyClass:
    pass

Another place pass can be used is as a place-holder for a function or conditional body when you are working on new code, allowing you to keep thinking at a more abstract level. The pass is silently ignored:

In [140]:
def initlog(*args):
    pass   # Remember to implement this!

---

<a id="section_markdown"></a>
## **Markdown Language**

### **Headings** 
Use the number sign (#) followed by a blank space for notebook titles and section headings:

* # Heading 1
* ## Heading 2
* ### Heading 3
* #### Heading 4
* ##### Heading 5
* ###### Heading 6
* ####### Heading 7

### **Emphasis**
Use the following code to emphasize text:  
* Bold text: __string__ or **string**
* Italic text: _string_ or *string*
* Bold Italic text: ___string___ or ***string***

* You can strike through text using HTML:
<s>this is strike through text</s>

### **Colors**
Colors: Use this code: `<font color=blue|red|green|pink|yellow>Text</font>`.

* <font color=blue>Blue text.</font>
* <font color=red>Red text.</font>


### **Mathematical symbols**
Surround mathematical symbols with a dollar sign (\\$ \\$), for example:  
$p(x)=\frac{1}{\sigma \sqrt{2\pi}} e^{-\frac{1}{2} \frac{(x-\mu)^2}{\sigma^2} }$.  
The symbols should be written in Latex. Please refer to [A (Not So) Short Introduction to $\LaTeX 2\varepsilon$](https://ctan.org/tex-archive/info/lshort/english/?lang=en).


### **Monospace font**
Surround text with a grave accent (\`) also called a back single quotation mark, for example:
Clive Humby said `Data is the new oil`.  

You can use the monospace font for file paths, file names, message text that users see, or text that users enter.

### **Line breaks**
Sometimes markdown doesn’t make line breaks when you want them. To force a linebreak, use \<br> or double space.


### **Indenting and Block Quoting**
Use the greater than sign (>), for example:  
>Text that will be indented when the Markdown is rendered.
>>Text that will be indented when the Markdown is rendered.

Any subsequent text is indented until the next carriage return.

### **Including Code Examples**

If you want to signify that a particular section of text is actually an example of code, you can use backquotes to surround the code example. These will switch the font to monospace, which creates a clear visual formatting difference between the text that is meant to be code and the text that isn't.

Code can either in the middle of a paragraph, or as a block. Use a single backquote to start and stop code in the middle of a paragraph. Here's an example:

The word `monospace` will appear in a code-like form.

***Note:*** If you want to include a literal backquote in your code example you must suround the whole text block in double backquotes like this:

`` Look at this literal backquote ` ``

To include a complete code-block inside a Markdown cell, use triple backquotes. Optionally, you can put the name of the language that you are quoting after the starting triple backquotes, like this:

```python
def function(n):
    return n + 1
```


### **Bullets**
To create a circular bullet point, use one of the following methods. Each bullet point must be on its own line.
- A hyphen (\-) followed by one or two spaces, for example: \- Bulleted item
+ A plus (\+) and a space, for example:\+ Bulleted item
* An asterisk (\*) followed by one or two spaces, for example: \* Bulleted item

To create a sub bullet, press Tab or space(s) before entering the bullet point using one of the methods described above. For example:  
- Main bullet point
    - Sub bullet point
    - Sub bullet point

### *Numbered lists*
To create a numbered list, enter 1. followed by a space, for example:  
1. Numbered item
1. Numbered item

For simplicity, you use 1. before each entry. The list will be numbered correctly when you run the cell.

To create a substep, press Tab or space(s) before entering the numbered item, for example:  
1. Numbered item 1
    1. Numbered item 1.1
    2. Numbered item 1.2
2. Numbered item 2

### **Colored note boxes**
Use one of the following <div> tags to display text in a colored box.  

Blue boxes (alert-info):  
<div class="alert alert-block alert-info">
<b>Tip:</b> Use blue boxes (alert-info) for tips and notes. 
If it’s a note, you don’t have to include the word “Note”.
</div>

Yellow boxes (alert-warning):  
<div class="alert alert-block alert-warning">
<b>Example:</b> Use yellow boxes for examples that are not 
inside code cells, or use for mathematical formulas if needed.
</div>

Green boxes (alert-success):
<div class="alert alert-block alert-success">
<b>Up to you:</b> Use green boxes sparingly, and only for some specific 
purpose that the other boxes can't cover. For example, if you have a lot 
of related content to link to, maybe you decide to use green boxes for 
related links from each section of a notebook.
</div>
    
Red boxes (alert-danger):
<div class="alert alert-block alert-danger">
<b>Just don't:</b> In general, avoid the red boxes. These should only be
used for actions that might cause data loss or another major issue.
</div>


### **Graphics**
You can attach image files directly to a notebook in Markdown cells by dragging and dropping it into the cell.  
To add images to other types of cells, you must use a graphic that is hosted on the web and use the following code to insert the graphic:

![Python 3](figs/python-logo.png)

<img src="figs/python-logo.png" alt="Python 3 logo" title="Python 3" />

### **Insert Symbols**
Use &# followed by the decimal or xHEX reference number for the shape and semicolon, for example:

ASCII:  
&#64; &#34; &#x22; &quot;  
See [HTML ASCII Reference](https://www.w3schools.com/charsets/ref_html_ascii.asp).

UNICODE UTF-8:  
&euro; &#8364; &#x20AC;  
&#9639; &#x25ED;  
&#x2600; &#x2602; &#x26C5;  
See [HTML Unicode (UTF-8) Reference](https://www.w3schools.com/charsets/ref_html_utf8.asp).


### **Other Characters**
 
* Ampersand &amp; Ampersand

* &lt; angle brackets &gt;

* &quot; quotes &quot;

### **Horizontal Lines and Section Breaks**
On a new line, enter three asterisks:  
***
This is a Python course. Draw line with \***.
***

____
This is a Python course. Draw line with \___.
____


------
This is a Python course. Draw line with \------.

------

---
This is a Python course. Draw line with \---.

---

* * * 
This is a Python course. Draw line with \* * *.
* * * 

_ _ _ 
This is a Python course. Draw line with \_ _ _.
_ _ _ 

- - - 
This is a Python course. Draw line with \- - -.
- - - 



### ***Backslash Escape***
What happens if you want to include a literal character, like a \#, that usually has a specific function in Markdown? Backslash Escape is a function that prevents Markdown from interpreting a character as an instruction, rather than as the character itself. It works like this:

\# Wow, this isn't a header.  
\*** , this is not a section break.

Markdown allows you to use a backslash to escape from the functions of the following characters:

* \\ backslash
* \` backtick
* \* asterisk
* \_ underscore
* \{\} curly braces
* \[\] square brackets
* \(\) parentheses
* \# hashtag
* \+ plus sign|
* \- minus sign (hyphen)
* \. dot
* \! exclamation mark

### **Internal Link**

[Link to Markdown Language](#markdown_language)

You can add an ID above the section, and refer to this id.

[Markdown Language](#section_markdown)


### **External links**
To link to an external site, use the following code:
__[Ethan's Profile](https://sites.google.com/view/yifengli)__.<br>
Surround the link with two underscores (\_) on each side.

* Automatic Links: https://sites.google.com/view/yifengli
* Standard Links: [Ethan's Profile](https://sites.google.com/view/yifengli)
* Standard Links With Mouse-Over Titles: [Ethan's Profile](https://sites.google.com/view/yifengli "CS@Brock")

### **Tables**

In Markdown, you can make a table by using vertical bars and dashes to define the cell and header borders:  

|Header|Header|Header|Header|
|------|------|------|------|
|Cell  |Cell  |Cell  | Cell |
|Cell  |Cell  |Cell  | Cell |
|Cell  |Cell  |Cell  | Cell |
|Cell  |Cell  |Cell  | Cell |

Making a table this way might be especially useful if you want your document to be legible both rendered and unrendered. However, you don't need to include all of those dashes, vertical bars, and spaces for Markdown to understand that you're making a table. Here's the bare minimum you would need to create the table above:

Header|Header|Header|Header
-|-|-|-
Cell|Cell|Cell|Cell
Cell|Cell|Cell|Cell
Cell|Cell|Cell|Cell
Cell|Cell|Cell|Cell

It's important to note that the second line of dashes and vertical bars is essential. If you have just the line of headers and the second line of dashes and vertical bars, that's enough for Markdown to make a table.

Another important formatting issue has to do with the vertical bars that define the left and right edges of the table. If you include all the vertical bars on the far left and right of the table, like in the first example above, Markdown will ignore them completely. But, if you leave out some and include others, Markdown will interpret any extra vertical bar as an additional cell on the side that the bar appears in the unrendered version of the text. This also means that if you include the far left or right vertical bar in the second line of bars and dashes, you must include all of the otherwise optional vertical bars (like in the first example above).

Cell Justification: If not otherwise specified the text in each header and cell of a table will justify to the left. If, however, you wish to specify either right justification or centering, you may do so like this:

**Centered, Right-Justified, and Regular Cells and Headers**: 

centered header | regular header | right-justified header | centered header | regular header
:-:|-|-:|:-:|-
centered cell|regular cell|right-justified cell|centered cell|regular cell
centered cell|regular cell|right-justified cell|centered cell|regular cell

While it is difficult to see that the headers are differently justified from one another, this is just because the longest line of characters in any column defines the width of the headers and cells in that column.

**Note**: You cannot make tables directly beneath a line of text. You must put a blank line between the end of a paragraph and the beginning of a table.

### **References**

* [IBM]:https://www.ibm.com/support/knowledgecenter/en/SSGNPV_1.1.3/dsx/markd-jupyter.html
* [Jmanual]:https://jupyter.brynmawr.edu/services/public/dblank/Jupyter%20Notebook%20Users%20Manual.ipynb
* [HTML]:https://www.w3schools.com/html/default.asp